private/bufpkg/bufmodule/stats.go

```diff
@@ -0,0 +1,147 @@
+// Copyright 2020-2023 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bufmodule
+
+import (
+	"context"
+
+	"github.com/bufbuild/protocompile/ast"
+	"github.com/bufbuild/protocompile/parser"
+	"github.com/bufbuild/protocompile/reporter"
+	"go.uber.org/multierr"
+)
+
+// ComputeStats gathers some simple stats about the size of the given module.
+func ComputeStats(ctx context.Context, module Module) (*Stats, error) {
+	infos, err := module.SourceFileInfos(ctx)
+	if err != nil {
+		return nil, err
+	}
+	handler := reporter.NewHandler(reporter.NewReporter(
+		func(reporter.ErrorWithPos) error {
+			// never aborts
+			return nil
+		},
+		nil,
+	))
+	var errs error
+	stats := &Stats{}
+	for _, info := range infos {
+		file, err := module.GetModuleFile(ctx, info.Path())
+		if err != nil {
+			errs = multierr.Append(errs, err)
+		}
+		func() {
+			defer func() {
+				errs = multierr.Append(errs, file.Close())
+			}()
+			astRoot, err := parser.Parse(info.Path(), file, handler)
+			if err != nil {
+				stats.NumFilesWithSyntaxErrors++
```

Comment:
Should this return instead of passing a `nil` file node?

Reply:
Updated. In most cases, `parser.Parse` returns an error and a _non-nil_ file node. So, yes, we still want to examine the AST in that case. But I did add a check to see if the AST is nil, to prevent nil-pointer panics.